### ISSUE TITLE
test: remove monolithic mock files (mock.h/cc)

### ISSUE THREAD

**Issue by mattklein123** [envoyproxy/envoy#8770]
I have not analyzed this in detail, but I'm fairly certain that one of the reasons most of our tests take so much RAM to compile (and thus sometimes limit CPU throughput for compilation) are the giant monolithic mock files that we have for each namespace. 

This is a truly heinous tech debt task, but I think we need to kill the monolithic files and replace them with a h/cc per mock which would allow much more granular inclusion of mocks to compile.

**Comment by mattklein123** [comment]
cc @lizan for comment, I know this has come up before as a discussion point.

**Comment by zuercher** [comment]
I'm not sure if it would actually be less painful, but it might be possible to write a script to do this. 

https://eli.thegreenplace.net/2011/07/03/parsing-c-in-python-with-clang

**Comment by lizan** [comment]
I didn't analyzed though I guess it might result longer compilation because of the preprocessor will have to process header files many more times. But for sure it helps incremental build.

**Comment by wrowe** [comment]
> I didn't analyzed though I guess it might result longer compilation because of the preprocessor will have to process header files many more times. But for sure it helps incremental build.

I don't believe we retain any advantage of precompiled headers follow Google style. Because the first include file is often the .h file of the corresponding source, the headers and order of inclusion vary between most every generated object. Other sources don't include that by reference until much later in the include list.

**Comment by sunjayBhatia** [comment]
Here is a branch with some WIP to show how much improvement (at least on Windows) we get from breaking down one of the largest mock libraries (`//test/mocks/server:server_mocks`): https://github.com/greenhouse-org/envoy/tree/break-up-test-server-mocks

We only attempted to break down that one mock library and in cases where tests do not need most of the mocks included in that large library, there is some significant improvement (~16%). Otherwise, when including large pieces of the mocks (e.g.`MockFactoryContext`), there is not much improvement (~6%).

This was all manually done so far and we only tried to work with the `server_mocks` library, so there could definitely be more improvement to the process of breaking these monoliths down and different strategies taken (that could even include breaking up interfaces/impls into smaller pieces potentially).
